src/aliceVision/feature/sift/SIFT.cpp

```diff
@@ -11,18 +11,104 @@ namespace feature {
 
 int VLFeatInstance::nbInstances = 0;
 
+void SiftParams::setPreset(ConfigurationPreset preset)
+{
+    switch(preset.descPreset)
+    {
+        case EImageDescriberPreset::LOW:
+        {
+            _maxTotalKeypoints = 5000;
+            _peakThreshold = 0.02f;
+            break;
+        }
+        case EImageDescriberPreset::MEDIUM:
+        {
+            _maxTotalKeypoints = 10000;
+            _peakThreshold = 0.01f;
+            break;
+        }
+        case EImageDescriberPreset::NORMAL:
+        {
+            _maxTotalKeypoints = 20000;
+            _peakThreshold = 0.005f;
+            break;
+        }
+        case EImageDescriberPreset::HIGH:
+        {
+            _maxTotalKeypoints = 50000;
+            _peakThreshold = 0.001f;
+            break;
+        }
+        case EImageDescriberPreset::ULTRA:
+        {
+            _maxTotalKeypoints = 100000;
+            _peakThreshold = 0.0001f;
+            break;
+        }
+        default:
+            throw std::out_of_range("Invalid image describer preset enum");
+    }
+    if(preset.maxNbFeatures > 0)
+    {
+        _maxTotalKeypoints = preset.maxNbFeatures;
+    }
+    _firstOctave = 0;
+    _numScales = 3;
+    switch(preset.quality)
+    {
+        case EFeatureQuality::LOW:
+        {
+            // Work on lower image resolution
+            _firstOctave = 1;
+            break;
+        }
+        case EFeatureQuality::MEDIUM:
+        {
+            break;
+        }
+        case EFeatureQuality::NORMAL:
+        {
+            break;
+        }
+        case EFeatureQuality::HIGH:
+        {
+            // Increase scale space sampling to improve feature repeatability.
+            // "An analysis of scale-space sampling in SIFT, Ives Rey-Otero, Jean-Michel Morel, Mauricio Delbracio
+            // https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.703.9294&rep=rep1&type=pdf
+            _numScales = 6;
+            break;
+        }
+        case EFeatureQuality::ULTRA:
+        {
+            _numScales = 6;
+            break;
+        }
+    }
+    if(!preset.gridFiltering)
+    {
+        _gridSize = 0;
+    }
+    _contrastFiltering = preset.contrastFiltering;
+}
+
+
 std::size_t getMemoryConsumptionVLFeat(std::size_t width, std::size_t height, const SiftParams& params)
 {
   double scaleFactor = 1.0;
-  if(params._firstOctave > 0)
-      scaleFactor = 1.0 / std::pow(2.0, params._firstOctave);
-  else if(params._firstOctave < 0)
-      scaleFactor = std::pow(2.0, std::abs(params._firstOctave));
+
+  // if image resolution is low, increase resolution for extraction
+  const int firstOctave = params._firstOctave - (width*height < 5000*4000 ? 1 : 0);
```

Comment:
would be better to refactor this as the condition is used several times in the code, better to have it in only onw place for maintenability